### ISSUE TITLE
Implement the 128-bit floating point number type

### ISSUE THREAD

**Issue by NeilGirdhar** [numpy/numpy#7647]
Boost does it, could we just depend on Boost's implementation?  http://www.boost.org/doc/libs/1_58_0/libs/multiprecision/doc/html/boost_multiprecision/tut/floats/float128.html

**Comment by charris** [comment]
Best to keep the issues together rather than open a new one. We can borrow from boost but can't use c++, so some work would be needed, especially for the Python long type which is not stored as a common integer type. We could always go through a string, but that seems a bit unwieldy.

**Comment by NeilGirdhar** [comment]
Which issue is this a duplicate of?

**Comment by NeilGirdhar** [comment]
(https://github.com/numpy/numpy/issues/5272 just says to rename float128 to be float80_96 or float80_128 depending on the platform, which is different from implementing float128).

**Comment by charris** [comment]
If you mean true ieee quad precision, there should be something somewhere. Note that it is not generally supported by C, which has only the long double type, so it gets more complicated, especially if you want to support trig functions and such. There are some mailing list posts if that is what you mean, but adding the quad type would be some significant work. Note that gcc provide a library for that.

**Comment by NeilGirdhar** [comment]
To be honest, I'm not really interested in this feature.  I only created the bug because I couldn't find it.

**Comment by mhvk** [comment]
I think it would be very worthwhile to have a proper `float128`, so having an open issue to track that seems appropriate.

**Comment by mhvk** [comment]
As I have been implementing yet another class using a hack with two doubles, thought I'd ping here. It seems there is an IEEE standard for proper `float128` and gcc supports it on what seems a very complete list of processors [1].

It seems fine if this would just replace the current `float128`.

[1] https://www.gnu.org/software/libc/manual/html_node/Mathematics.html

**Comment by rgommers** [comment]
What about other compilers? A GCC-specific change will just make the confusion worse I'm afraid.

**Comment by seberg** [comment]
Maybe we should deprecate the float128 name? If users switch to `np.longdouble` instead, the name clears up for proper 128bit IEEE floating point numbers.

**Comment by NeilGirdhar** [comment]
`float128` and `float96` seem much clearer than `longdouble'.

**Comment by seberg** [comment]
@NeilGirdhar  yes, but they are also completely misleading and ill defined. `longdouble` at least is clear about mapping to the C long double definition.

**Comment by mhvk** [comment]
@rgommers - some googling suggests both the llwm and the intel compiler also support it, though the latter perhaps not trivially (but presumably we can get some help from the intel folks). Are there other compilers are you thinking about?

Since currently `np.float128` is not guaranteed to exist either, I'm not sure the situation will be worse than it is now. At least, if it exists, it will have consistent behaviour...

**Comment by mhvk** [comment]
Deprecating `float128` in favour of `longdouble` would seem to make sense - as noted, it at least has the benefit of not suggesting what precision or storage requirement the type has. Though there still is a question of how to represent it in a `dtype`.

But the good news is that while we may cause performance loss for the rare users that use the current `longdouble`, we cannot cause precision regressions.

**Comment by NeilGirdhar** [comment]
@seberg  I understand the draw of having this match C's definition of long double, but this is going to cause problems when GPUs start supporting float128.  I like the suggestion in #5272 to rename `float128` to be `float80_96` or `float80_128` depending on the platform, and then make `float128` the true quadruple precision float.

**Comment by seberg** [comment]
@NeilGirdhar I was merely thinking we could try to Deprecate or if we already know we can pull of true float128 now Future warn float128, naming it differently, such as `float80_96` would seem good (although I am not sure that is correct on Macs? Weren't they double-double users?).

**Comment by rgommers** [comment]
> @rgommers - some googling suggests both the llwm and the intel compiler also support it, though the latter perhaps not trivially (but presumably we can get some help from the intel folks). Are there other compilers are you thinking about?

MSVC mainly.

**Comment by mhvk** [comment]
Hmm, MSVC status seems rather unclear for quad precision. Though I'm not sure it is better to have a regular double inside a `float128`, as I gather is the case now according to the [numpy docs](https://docs.scipy.org/doc/numpy/user/basics.types.html). cc @eric-wieser in case he knows.

**Comment by matthew-brett** [comment]
Also - there will be users who have saved e.g. npy files with float128 / float96 type, where float128 is in fact padded float80.  If you re-purpose float128, you'll break those files.

**Comment by NeilGirdhar** [comment]
@matthew-brett Can't you just bump the npy version number?

**Comment by matthew-brett** [comment]
I guess - maybe with some extra code to cast the loaded data to the new types, say `float80_128`.  This just a marker to remind us that we will need those types, and to do those changes.

**Comment by ngoldbaum** [comment]
For what it's worth, numpy-quaddtype exists now: https://numpy.org/numpy-user-dtypes/quaddtype/
